src/tools/replay/trace_replay.ml

```diff
@@ -651,30 +651,76 @@ struct
     let close () = Lwt.return (Ok ())
   end
 
+  module Commit_stats = struct
+    let h = open_out "/tmp/tezos_replay_stats"
+
+    type t = { mutable previous_timer : float; mutable times : float list }
+
+    let t = { previous_timer = Unix.gettimeofday (); times = [] }
+    let steps = 30.0
+    let isteps = int_of_float steps
+    let time0 = isteps * int_of_float (t.previous_timer /. steps)
+
+    let print_stats () : unit =
+      let arr = Array.of_list t.times in
+      t.times <- [];
+      Array.sort Float.compare arr;
+      let len = Array.length arr in
+      let q0 = arr.(0) in
+      let q1 = arr.(len / 4) in
+      let q2 = arr.(len / 2) in
+      let q3 = arr.(len * 3 / 4) in
+      let q4 = arr.(len - 1) in
+      let time = isteps * int_of_float (t.previous_timer /. steps) in
+      Printf.fprintf h "%i\t%i\t%f\t%f\t%f\t%f\t%f\n%!" (time - time0) len q0 q1
+        q2 q3 q4
+
+    let add () =
+      let now = Unix.gettimeofday () in
+      if int_of_float (t.previous_timer /. steps) <> int_of_float (now /. steps)
+      then print_stats ();
+      t.times <- (now -. t.previous_timer) :: t.times;
+      t.previous_timer <- now
+  end
+
   let exec_commit rs ((time, message, c), hash) =
     let level = rs.current_row.level in
     Stat_recorder.set_stat_specs (specs_of_row rs.current_row);
     let time = Time.Protocol.of_seconds time in
     let c = on_lhs_context rs c in
     let* hash' = Context.commit ~time ?message c in
     on_rhs_hash rs hash hash';
-    Stdlib.Hashtbl.add rs.hash_per_level level hash';
+    Stdlib.Hashtbl.add rs.hash_per_level rs.current_block_idx hash';
```

Comment:
(For context, this line change is a hack to allow the selection of the GC commit at the previous `split`, as it turns out that the cycles can get misaligned otherwise... I'm only mentioning it in case it interfere with @adatario trace update, just so you know there's a trap there :P )

Reply:
I think we may have fallen into the trap. With this patch applied, I see the following error when 8192 commits have been processed (which is the value of `--gc-when`):

```
Replaying trace   8192/145000 commits 09:23 (ETA: --:--) [---------------]   5%
replay: internal error, uncaught exception:
        File "src/trace/stats_trace_recorder.ml", line 927, characters 14-20: Assertion failed
        Raised at Lwt.Miscellaneous.poll in file "src/core/lwt.ml", line 3077, characters 20-29
        Called from Lwt_main.run.run_loop in file "src/unix/lwt_main.ml", line 31, characters 10-20
        Called from Lwt_main.run in file "src/unix/lwt_main.ml", line 118, characters 8-13
        Re-raised at Lwt_main.run in file "src/unix/lwt_main.ml", line 124, characters 4-13
        Called from Cmdliner_term.app.(fun) in file "cmdliner_term.ml", line 24, characters 19-24
        Called from Cmdliner_eval.run_parser in file "cmdliner_eval.ml", line 34, characters 37-44
```

Reply:
Hmm I've not seen this assert before hmhm any chance it could be related to the `skip-gc` perhaps? (since the assert seems unhappy that there was no GC?)

Reply:
Ha ok yeah you're right, it's also possible that it doesn't find a target commit the first time the GC could run due to a +-1 offset? (I've no idea why it still tries to reach for the latest GC stats though)

Reply:
I removed the `skip-gc` part of the original patch and pushed it as a separate [branch/commit](7358ff31ea8730aa329bdc101a7b64ca12052263). @adatario can look at this, when overhauling the GC triggering code.  I'll report if I see any more failures on the replay benchmark run while dumping these stats.

Reply:
Ui, it's all a bit of a mine-field :)

I also think that this is an interference with `skip-gc` and the new recording of gc events.